pkg/controller/release/strategy_patches.go

```diff
@@ -0,0 +1,97 @@
+package release
+
+import (
+	"encoding/json"
+
+	"k8s.io/apimachinery/pkg/api/equality"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+
+	shipper "github.com/bookingcom/shipper/pkg/apis/shipper/v1alpha1"
+)
+
+type StrategyPatch interface {
+	PatchSpec() (string, schema.GroupVersionKind, []byte)
+	Alters(interface{}) bool
+	IsEmpty() bool
+}
+
+type CapacityTargetSpecPatch struct {
+	Name    string
+	NewSpec *shipper.CapacityTargetSpec
+}
+
+var _ StrategyPatch = (*CapacityTargetSpecPatch)(nil)
+
+func (c *CapacityTargetSpecPatch) PatchSpec() (string, schema.GroupVersionKind, []byte) {
+	patch := make(map[string]interface{})
+	patch["spec"] = c.NewSpec
+	b, _ := json.Marshal(patch)
+	return c.Name, shipper.SchemeGroupVersion.WithKind("CapacityTarget"), b
+}
+
+func (c *CapacityTargetSpecPatch) Alters(o interface{}) bool {
+	ct, ok := o.(*shipper.CapacityTarget)
+	if !ok || c.IsEmpty() {
+		return false
+	}
+	return !equality.Semantic.DeepEqual(&ct.Spec, c.NewSpec)
+}
+
+func (c *CapacityTargetSpecPatch) IsEmpty() bool {
+	return c == nil || c.NewSpec == nil
+}
+
+type TrafficTargetSpecPatch struct {
+	Name    string
+	NewSpec *shipper.TrafficTargetSpec
+}
+
+var _ StrategyPatch = (*TrafficTargetSpecPatch)(nil)
+
+func (c *TrafficTargetSpecPatch) PatchSpec() (string, schema.GroupVersionKind, []byte) {
+	patch := make(map[string]interface{})
+	patch["spec"] = c.NewSpec
+	b, _ := json.Marshal(patch)
+	return c.Name, shipper.SchemeGroupVersion.WithKind("TrafficTarget"), b
+}
+
+func (c *TrafficTargetSpecPatch) Alters(o interface{}) bool {
+	tt, ok := o.(*shipper.TrafficTarget)
+	if !ok || c.IsEmpty() {
+		return false
+	}
+	return !equality.Semantic.DeepEqual(&tt.Spec, c.NewSpec)
```

Comment:
Given how `checkTraffic` works, I don't think this `DeepEqual` check will ever return true, so I'm skeptical about having it here, both for the small performance penalty, and for the possibility of this hiding bugs in `checkTraffic` itself if it ever starts spewing out no-op patches. The same goes for `CapacityTargetSpecPatch`.